azure/services/async/async.go

```diff
@@ -106,20 +106,23 @@ func (s *Service[C, D]) CreateOrUpdateResource(ctx context.Context, spec azure.R
 
 	result, poller, err := s.Creator.CreateOrUpdateAsync(ctx, spec, resumeToken, parameters)
 	errWrapped := errors.Wrapf(err, "failed to create or update resource %s/%s (service: %s)", rgName, resourceName, serviceName)
-	if poller != nil {
+	if poller != nil && azure.IsContextDeadlineExceededOrCanceledError(err) {
 		future, err := converters.PollerToFuture(poller, infrav1.PutFuture, serviceName, resourceName, rgName)
 		if err != nil {
 			return nil, errWrapped
 		}
 		s.Scope.SetLongRunningOperationState(future)
```

Comment:
The function change here is to only `SetLongRunningOperationState` when the error is something like "not done yet." In all other cases, successful or not, we `DeleteLongRunningOperationState`.

Reply:
Asking for clarity(I may be way off the workflow), what happens to _LongRunningRequests_ if they fail or if the higher abstraction resource cancels the context due to some failure ? Do we intend to re-queue this resource's request and not save its state in `Status.Condition`?

Reply:
Right. If an operation fails, then it's done, and we shouldn't keep polling that specific operation because it will remain failed. In that case we return the error which will requeue the resource and a new operation will start.